### ISSUE TITLE
Feature request: ignore third party libraries.

### ISSUE THREAD

**Issue by mrice32** [trailofbits/slither#119]
It would be nice to be able to ask slither to ignore third party libraries (or just certain paths) when running in a truffle directory. For instance, open zeppelin's libraries are triggering some warnings (the external function and naming convention detectors) in our codebase. Rather than completely disable these detectors in our CI, we would like to be able to ignore warnings that point to open zeppelin source files. If you think this might be a reasonable feature to add, I may take a stab at implementing it.

**Comment by montyly** [comment]
Hi @mrice32 , that's a really good idea!

I am not entirely sure how to implement it correctly though.

It's easy to filter the json results according to a blacklist of source mapping directories, but it would be a bit more complex to filter the results printed to the standard output based on a list of directories (because of how the detectors work).

An alternative solution is to allow Slither to 'load' previous results (using the json), and to print a finding only if it was not already reported. Though this may lead a user to 'miss' a finding if all the previously reported bugs are discarded. 
Maybe we could provide an interactive command line that allows to 'validate' what results should not be reported in the future? So the first time you run slither, you can, one by one, discard the bugs that should not be reported for the next run.

If you have any suggestions, I would be happy to hear them.

**Comment by samparsky** [comment]
@montyly Since slither contains the source path of a file, we can check the source path for the root dir of third-party libraries and ignore it from being printed.

For example if we have flag to ignore like
`slither --ignorefile openzeppelin-solidity,demo`

We can pass the list of root dirs during the initialisation of the `AbstractDetector` and in the `log` function we can check the source path of the solidity file and see if it contains the root dir `openzeppelin-solidity or demo` and ignore it from being printed.

This is just off the top of my head I might have to dig a bit deeper.

**Comment by montyly** [comment]
A bug can be related to something in `openzeppelin-solidity`, but not a bug of the library itself.
For example, if you shadow a variable of the library, the path will point `openzeppelin-solidity`. So we can't just filter the log information based on a list of paths.

I am considering to add a function `is_valid_result` to `AbstractDetector`, that would take a json result and check if it be should be saved/printed. 

A detector like
https://github.com/trailofbits/slither/blob/eae46813ab0b07e6d6cf52e9995510ead44f1c03/slither/detectors/examples/backdoor.py#L22-L29

Would become
```python 
 # Add the result in result 
 json = self.generate_json_result(info) 
 self.add_function_to_json(f, json) 
 if self.is_valid_result(json):
   results.append(json)

   info = 'Backdoor function found in {}.{} ({})\n' 
   info = info.format(contract.name, f.name, f.source_mapping_str) 
   # Print the info 
   self.log(info) 
```

We can easily ignore third-party library using `is_valid_result` and add a 'load' previous results feature.

I will see if this pattern works for all the detectors. We probably need to abstract this in sub functions, otherwise the code will become too complex for new contributors.

**Comment by montyly** [comment]
Slither can now:
- filter paths using [`--filter-paths path1,path2`](https://github.com/trailofbits/slither/wiki/Usage#path-filtering)
- Save results to not show them during next runs using [`--triage-mode`](https://github.com/trailofbits/slither/wiki/Usage#triage-mode)

These features are merged in `dev`, and will be available in the upcoming 0.6.0 release.

**Comment by mrice32** [comment]
That's awesome!! Thanks @montyly!
